File: Java/Euler (Java)/src/problems/P026xP050/P035.java
package problems.P026xP050;

import java.util.List;

import util.Primes;
import base.Problem;

/**
 * P035<br>
 * The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.
 * There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.
 * How many circular primes are there below one million?
 * Ans: 55 
 * @return
 */
public class P035 extends Problem
{
   public P035()
   {
      problemNumber = 35;
      variations = 2;
   }

   @Override
   protected String execute(int variation)
   {
      switch (variation)
      {
         case 0:
            return bruteForce() + "";
         case -1:
         case 1:
            return faster() + "";
      }
      return null;
   }

   /**
    * I think we can get the primes in an efficient way and then just test them all
    * @return
    */
   private long bruteForce()
   {
   	int total = 0;
   	List<Integer> primes = Primes.EratosthenesPrimesUpTo(1000000);
   	for (int p : primes)
   	{
   		int next = rotateRight(p);
   		boolean isCircular = true;
   		while (next != p)
   		{
   			if (!Primes.isPrime(next))
   			{
   				isCircular = false;
   				break;
   			}
   			next = rotateRight(next);
   		}
   		if (isCircular)
   		{
   			total++;
   		}
   	}
   	return total;
   }
   
   private int rotateRight(int num)
	{
		int i = num % 10;
		int j = (int) (num/10);
		int size = 1;
		while (j >= size)
			size *= 10;
		j += size * i;
		return j;
	}
   
   /**
    * Keeping track of which primes are good and bad didn't work well. Took almost a full 
    * second instead of 40ms. So instead, let's look at the digits. Ignore anything that
    * has an even number or a 5 in it
    * @return
    */
   private long faster()
   {
   	int total = 0;
   	List<Integer> primes = Primes.EratosthenesPrimesUpTo(1000000);
   	for (int p : primes)
   	{
   		if (p >= 10)
   		{
   			if (!checkDigits(p))
   				continue;
   		}
   		int next = rotateRight(p);
   		boolean isCircular = true;
   		while (next != p)
   		{
   			if (!Primes.isPrime(next))
   			{
   				isCircular = false;
   				break;
   			}
   			next = rotateRight(next);
   		}
   		if (isCircular)
   		{
   			total++;
   		}
   	}
   	return total;
   }
   
   private boolean checkDigits(int p)
   {
   	while (p != 0)
   	{
   		int n = p % 10;
   		if (n % 2 == 0 || n % 5 == 0)
   			return false;
   		p = p / 10;
   	}
   	return true;
   }
}

File: OpenCL/Euler (OpenCL)/src/problems/P031xP040/P035.cpp
#include "../../base/problem.h"

#include <set>

class P035: public Problem
{
	/**
	 * P035<br>
	 * The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.
	 * There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.
	 * How many circular primes are there below one million?
	 * Ans: 55
	 */
	public:
		P035()
		{
			problem_number = 35;
			variations = 2;
		}
		~P035() {}

		bool setup(int variation)
		{
			bool retval = false;
			cl = new CLKernel(kernelsize);
			switch (variation)
			{
				case 0:
					retval = brute_setup();
					break;
				case 1:
				case -1:
					retval = opt_setup();
					break;
				default:
					retval = false;
			}
			return retval;
		}

		bool teardown(int variation)
		{
			bool retval = false;
			switch (variation)
			{
				case 0:
					retval = brute_teardown();
					break;
				case 1:
				case -1:
					retval =opt_teardown();
					break;
				default:
					retval = false;
			}
			cl->releaseKernel();
			return retval;
		}

		void execute(int variation)
		{
			switch (variation)
			{
				case 0:
					brute();
					break;
				case 1:
				case -1:
					optimized();
					break;
				default:
					std::cout << "NOOP!" << std::endl;
					break;
			}
		}

	private:

		int size;

		/**
		 * Brute force version
		 *
		 */
		bool brute_setup()
		{
			size = 1000000;
			int initialVal = 1; // 2

			cl->init("src/problems/P031xP040/P035.cl","brute");
         cl->createReadWriteMemory( sizeof(int),&answer_mem);
         cl->transferCPUObjectToGPU(sizeof(int),&answer_mem,&initialVal);
         cl->setupOpenCLKernelArgs(2,
         		sizeof(cl_mem), &answer_mem,
					sizeof(int), &size);
			cl->waitForQueue(); // sync
			setThreads(size);
			return true;
		}

		bool brute_teardown()
		{
			int ans;
			if (!cl->transferGPUObjectToCPU(sizeof(int), &answer_mem, &ans))
			{
				std::cout << "ERROR:: Could not read answer from GPU!" << std::endl;
				return false;
			}

			answer = to_string(ans);
			clReleaseMemObject(answer_mem);
			return true;
		}

		int brute()
		{
			cl->launchKernel(threadsize);
			cl->waitForQueue();
			return 0;
		}

		/**
		 *
		 */
		bool opt_setup()
		{
			size = 1000000;
			int initialVal = 2; // 2 and 5

			cl->init("src/problems/P031xP040/P035.cl","optimized");
			cl->createReadWriteMemory( sizeof(int),&answer_mem);
			cl->transferCPUObjectToGPU(sizeof(int),&answer_mem,&initialVal);
			cl->setupOpenCLKernelArgs(2,
					sizeof(cl_mem), &answer_mem,
					sizeof(int), &size);
			cl->waitForQueue(); // sync
			setThreads(size);
			return true;
			return true;
		}

		bool opt_teardown()
		{
			int ans;
			if (!cl->transferGPUObjectToCPU(sizeof(int), &answer_mem, &ans))
			{
				std::cout << "ERROR:: Could not read answer from GPU!" << std::endl;
				return false;
			}

			answer = to_string(ans);
			clReleaseMemObject(answer_mem);
			return true;
		}

		int optimized()
		{
			cl->launchKernel(threadsize);
			cl->waitForQueue();
			return 0;
		}
};
